service/src/main/kotlin/fi/espoo/evaka/messaging/MessageController.kt

```diff
@@ -488,6 +498,30 @@ class MessageController(
             .also { Audit.MessagingMessageReceiversRead.log(meta = mapOf("count" to it.size)) }
     }
 
+    @PostMapping("/{accountId}/undo-message")
+    fun undoMessage(
+        db: Database,
+        user: AuthenticatedUser,
+        clock: EvakaClock,
+        @PathVariable accountId: MessageAccountId,
+        @RequestParam messageId: MessageId?,
+        @RequestParam contentId: MessageContentId?
+    ): MessageDraftId? {
+        return db.connect { dbc ->
+                requireMessageAccountAccess(dbc, user, clock, accountId)
+                dbc.transaction {
+                    when {
+                        messageId != null ->
+                            it.undoMessageReply(clock.now(), accountId, messageId).let { null }
+                        contentId != null ->
+                            it.undoMessageAndAllThreads(clock.now(), accountId, contentId)
+                        else -> throw BadRequest("Either messageId or contentId is required")
```

Comment:
It would be nice to adjust typings in a way that you can't give both `messageId` and `contentId`, or add separate routes for canceling a reply or a new thread, or at least add runtime validation checking that both are not passed.

Reply:
So which solution would be preferable in your opinion? Separate routes? I don't have strong opinions on any of the options.

Reply:
I've split the route into two.